[Eyo Wei Chin] iP

EyoWeiChin opened the pull request.

EyoWeiChin committed:
Update Duke.java

Add a skeletal version that simply greets the user and exits.

EyoWeiChin committed:
Level 1: Greet, Echo, Exit

EyoWeiChin committed:
Level 2: Add, List

Add function to store user input and display them in a list when requested

EyoWeiChin committed:
Level 3: Mark as Done

Add the ability to mark tasks as done.

EyoWeiChin committed:
Update Duke.java to handle out of bounds user input

Fix a case where user enters 'done' without specifying a task

EyoWeiChin committed:
Update Duke.java and Task.java

Minor adjustments made to comply with coding standards

EyoWeiChin committed:
Level 4 Todos, Events, Deadlines

Add support to track three new types of tasks.

EyoWeiChin committed:
A-TextUiTesting

Add automated testing script
Add testcase and expected outcome

EyoWeiChin committed:
Refactor Duke.java

Improve code quality with refactoring, and introducing constants

EyoWeiChin committed:
Update Duke.java and the testcase

Refactor code by introducing message constants.
Update expected output to reflect the changes in some message

woozije commented on src/main/java/Duke.java:
```diff
@@ -1,10 +1,259 @@
+
+    /**
+     * Returns the split inputParts into 3 distinct data: Command, Task name, Task Parameters
+     * @param inputParts Separated user input into 2 parts by the first spacing
```
Should there be an empty line left in between the description and parameter section? I noticed the same issue in other places too. 
Perhaps you could also add a punctuation behind each parameter description.

EyoWeiChin replied:
Thank you for your review, I have resolved the issue!

woozije reviewed (commented):
In general, I like how the code complies to coding standards so that the code is easier to read.

mrwsy1 commented on src/main/java/Duke.java:
```diff
@@ -1,10 +1,259 @@
+import java.util.ArrayList;
+import java.util.Scanner;
+
 public class Duke {
```
nice code, very neat.

mrwsy1 reviewed (commented):
Looks good.

EyoWeiChin committed:
Update Duke and the testcase

Update Duke to catch array out of bounds error for input processing
Update testcase to include erroneous cases to trigger the exception

EyoWeiChin committed:
Add DukeException.java

DukeException class for future execeptions specific to Duke

EyoWeiChin committed:
Merge branch 'branch-Level-5'

EyoWeiChin committed:
Add duke and duke.task packages for better organisation

Update runtest.bat accordingly so automated testing still works

EyoWeiChin committed:
Merge branch 'branch-A-Packages'

EyoWeiChin committed:
Level-6: Delete

Add ability to delete tasks from the task list

EyoWeiChin committed:
Level-7: Save

Add automated saving of tasklist to duke.txt file
Add automated loading of tasklist from duke.txt file

EyoWeiChin committed:
Refactor Duke.java

Improve code quality of the Save and Load feature

EyoWeiChin committed:
Merge branch 'branch-Level-6'

EyoWeiChin committed:
Merge branch 'branch-Level-7'

# Conflicts:
#	src/main/java/duke/Duke.java

EyoWeiChin committed:
Update Duke.java to save after a task is deleted

EyoWeiChin committed:
Update Duke.java to use a non-global ArrayList

EyoWeiChin committed:
Update DukeException to handle Duke related errors

EyoWeiChin committed:
Refactor Duke.Java to utilise more OOP

Refactor all messages to Messages.java class
Refactor input processing to Parser class
Refactor save/load to SaveManager class
Refactor tasks list to TaskList class
Refactor user interfaces to TextUI class

EyoWeiChin committed:
Update SaveManager class to be more readable

EyoWeiChin committed:
Remove static utility classes

EyoWeiChin committed:
A-MoreOOP: Add Command class and relevant subclasses

Each command now has its own class

EyoWeiChin committed:
Level-8: Dates and Times

Update AddCommand to handle LocalDate
Update Tasks, Deadline and Event class to better handle LocalDate

EyoWeiChin committed:
Add command that lists task due on a specific date

EyoWeiChin committed:
Level-9: Find

Add a command that lets user find task with keyword

EyoWeiChin committed:
Merge pull request #1 from EyoWeiChin/branch-Level-8

Merge branch level 8

EyoWeiChin committed:
Merge branch 'master' into branch-Level-9

# Conflicts:
#	src/main/java/duke/parser/Parser.java

EyoWeiChin committed:
Merge pull request #2 from EyoWeiChin/branch-Level-9

Merge level-9: Find

Samuel787 commented on src/main/java/duke/commands/DoneCommand.java:
```diff
@@ -0,0 +1,40 @@
+        } else {
+            tasks.completeTask(taskIDToFinish);
+            saveManager.saveTaskList(tasks);
+            result = "'" + tasks.getTasks().get(taskIDToFinish).getTaskName().trim() + "'";
+            result += System.lineSeparator() + Messages.MESSAGE_TASK_COMPLETED;
+            if (Task.getTotalTasks() == 0) {
+                result += Messages.MESSAGE_ALL_TASK_COMPLETED;
+            } else {
+                result += Messages.MESSAGE_TASKS_LEFT + Task.getTotalTasks();
```
You may practise SLAP here by extracting this out into a helper method

Samuel787 commented on src/main/java/duke/common/Messages.java:
```diff
@@ -0,0 +1,37 @@
+    /**
+     * Declare Messages to be used by Duke here.
+     */
+    public static final String MESSAGE_NO_SUCH_TASK = "That Task does not exist!";
+    public static final String MESSAGE_NO_REMAINING_TASKS = "No remaining tasks";
+    public static final String MESSAGE_TASK_ALREADY_COMPLETED = "That Task has already " +
+            "been completed, but let's shoot it again";
+    public static final String MESSAGE_TASK_COMPLETED = "marked as completed, well done!";
+    public static final String MESSAGE_ALL_TASK_COMPLETED = "All Task Completed!";
+    public static final String MESSAGE_TASKS_LEFT = "You a total of tasks remaining: ";
+    public static final String MESSAGE_GOODBYE = "Bye! Hope to see you again soon!";
+    public static final String MESSAGE_GREETING = "Hello! I'm Duke!" + System.lineSeparator()
+                    + "What can I do for you?";
+    public static final String MESSAGE_INVALID_OPTION = "Invalid option, Try again!"
+                            + System.lineSeparator();
+    public static final String ERROR_MESSAGE_NO_INFO = "Please provide more information!";
+    public static final String MESSAGE_REMOVED_TASK = "Noted! I have removed the task:";
+    public static final String ERROR_CANNOT_WRITE = "Unable to write to file: ";
+    public static final String MESSAGE_SUCCESSFUL_LOAD = "Save file loaded! Added the following tasks:";
+    public static final String ERROR_DUPLICATE_SAVE = "Save file already exists!";
+    public static final String MESSAGE_CREATED_SAVE_FILE = "No save file detected! New file was created!";
```
Good use of constants for the messages.

Samuel787 commented on src/main/java/duke/parser/Parser.java:
```diff
@@ -0,0 +1,98 @@
+public class Parser {
+
+    //Constant variables for Command logic
+    protected static final String RAW_COMMAND_DELIMIT = " ";
+    protected static final int SPLIT_INPUT_LIMIT = 2;
+    protected static final String COMMAND_DELETE = "delete";
+    protected static final String DELIMIT_DELETE = "";
+    protected static final String COMMAND_TODO = "todo";
+    protected static final String DELIMIT_TODO = "";
+    protected static final String COMMAND_DONE = "done";
+    protected static final String DELIMIT_DONE = "";
+    protected static final String COMMAND_DEADLINE = "deadline";
+    protected static final String DELIMIT_DEADLINE = "/by";
+    protected static final String COMMAND_EVENT = "event";
+    protected static final String DELIMIT_EVENT = "/at";
+    protected static final String COMMAND_LIST = "list";
+    protected static final String COMMAND_BYE = "bye";
+    protected static final String COMMAND_FIND = "find";
+    protected static final String DELIMIT_FIND = "";
+    protected static final String COMMAND_DUE = "due";
+    protected static final String DELIMIT_DUE = "";
```
Good job extracting them out as constants

Samuel787 commented on src/main/java/duke/storage/SaveManager.java:
```diff
@@ -0,0 +1,142 @@
+            Files.createDirectories(Paths.get(String.valueOf(DATA_FOLDER)));
+            Files.createFile(Paths.get(String.valueOf(filePath)));
+            throw new DukeException(Messages.MESSAGE_CREATED_SAVE_FILE);
+        } catch (IOException fileExists) {
+            //The actual reading takes place here
+            savedFileReader(loadedTasks);
+        }
```
Exceptions are meant to gracefully prevent the program from failing and not to perform functional logic

Samuel787 reviewed (commented):
Generally, good coding standards adopted and code quality is good. Well done.

EyoWeiChin committed:
A-JavaDoc

Add Java doc comments to non-private classes and non-trivial methods

EyoWeiChin committed:
Refactor DoneCommand by extracting method out

Extract canTaskBeCompleted to follow SLAP

EyoWeiChin committed:
Update SaveManager to improve code quality

Update loadTaskList to not use exceptions as part of code logic
Update loadTaskList to use helper methods to follow SLAP

EyoWeiChin committed:
Merge branch 'master' into branch-A-JavaDoc

EyoWeiChin committed:
Merge pull request #3 from EyoWeiChin/branch-A-JavaDoc

Merge A-JavaDoc

EyoWeiChin committed:
Set theme jekyll-theme-leap-day

EyoWeiChin committed:
A-UserGuide: Add a User Guide

EyoWeiChin committed:
Refactor to improve code quality and coding standards

EyoWeiChin committed:
Add comments and Javadocs

EyoWeiChin committed:
Update README

Add more details for both README

EyoWeiChin committed:
Update JavaDocs to be more accurate and descriptive

EyoWeiChin committed:
Update Delete and Done to perform input validation

EyoWeiChin committed:
Add a FAQ and Command Summary

EyoWeiChin committed:
Update README to fix minor formatting issues

EyoWeiChin committed:
Update SaveManager class with minor fix

EyoWeiChin committed:
Add a simple Help Command

EyoWeiChin committed:
Merge branch 'branch-Help'

EyoWeiChin committed:
Update Add and Due to record time as well

EyoWeiChin committed:
Refactor to improve code quality

Refactor some regex and constant
Delete unused variables
Update HelpCommand with new information.

EyoWeiChin committed:
Update README and User Guide

Change README back to original form
Update User guide to reflect recent changes

EyoWeiChin committed:
Update JavaDocs of several components

Add clarity and missing information
